### ISSUE TITLE
UI::HtmlDialog bug in :use_content_size

### ISSUE THREAD

**Issue by dezmoHU** [SketchUp/api-issue-tracker#670]
# Bug Reports

1. Version: SketchUp 2021.1
2. OS Platform: W10

If you set the use_content_size in [UI::HtmlDialog ](https://ruby.sketchup.com/UI/HtmlDialog.html#initialize)Options Hash (properties)
```ruby
:use_content_size=> true, 
```
the **max_height** and **max_width** values are **swapped**!

See details here: 
[Sketchup Forum](https://forums.sketchup.com/t/use-content-size-true-swaps-the-max-height-and-max-width-htmldialog/170445/3?u=dezmo)

![use_true](https://user-images.githubusercontent.com/88205452/127688845-b08e1d86-6412-4dd0-9336-7e84176fc842.gif)

**Comment by DanRathbun** [comment]
A workaround is to use the new [`set_content_size`](https://ruby.sketchup.com/UI/HtmlDialog.html#set_content_size-instance_method) method rather than constructor arguments.

----

Also, some quick testing shows that the documentation **lacks** to inform us that the `use_content_size` flag also **affects the positioning**.

If you do not specify `top` or `left` (or you set them to `0`,) the window is positioned relative to the top left of the **content area** (not the upper left corner of the window frame,) and so will put the caption bar and left frame border offscreen.

This can make it difficult for users to know how to close the dialog unless it provides a close button in the content. (Users would need to know the built-in shortcut for closing the active window. Ie: `ALT+F4` on Windows, `Command+W` on Mac.)

I've had to come up with a position correction method to over come this bug:
```ruby
module Testing

  @dlg = nil unless defined?(@dlg)

  def self.cef_ver(content_sizing: true, center: false)

    return @dlg.bring_to_front if @dlg && @dlg.visible?

    @dlg = UI::HtmlDialog.new(
      dialog_title: "ChromeEmbeddedFramework",
      preferences_key: "CEF_VERSION",
      style: UI::HtmlDialog::STYLE_UTILITY,
      resizable:  true,
      scrollable: false,
      width:  848, # default frame width
      height: 453, # default frame height
      top:    0,
      left:   0,
      use_content_size: content_sizing # ignored <= SU2021.0
    )

    if content_sizing &&
    @dlg.respond_to?(:set_content_size) # SU2021.1+
      @dlg.set_content_size( 832, 414 )
    elsif @dlg.respond_to?(:set_size) # SU2017.0+
      @dlg.set_size( 848, 453 )
    end

    @dlg.center if center
    @dlg.set_url('chrome://version/')
    @dlg.show

    # Correct offscreen positioning:
    if @dlg.respond_to?(:get_position) && @dlg.visible?
      x, y = @dlg.get_position
      self.correct_position(x,y) if x <= 0 || y < 0
    end

  end ### cef_ver

  private

  def self.correct_position(x,y)
    return unless x <= 0 || y < 0
    puts "Position: x = #{x}, y = #{y}"
    puts "Correcting offscreen positioning ..."

    outer_width, outer_height = @dlg.get_size
    inner_width, inner_height = @dlg.get_content_size

    border = ((outer_width - inner_width)/2).round
    caption_bar = ((outer_height - inner_height) - border).round
    puts "border: #{border}"
    puts "caption bar: #{caption_bar}"
    # On Win10, always evals to an 8px border and 31px caption bar.

    # SU2021.1 the x origin is 8 pixels off towards the right (on Win10.)
    # May be due to MFC window metrics that no longer apply in Win10 with
    # it's single pixel borders (ie, Aero style is not used by default.)
    if Sketchup.platform == :platform_win
      x < 1 ? x1 = -8 : x1 = x - 8
    else
      x < 0 ? x1 = 0 : x1 = x
    end
    y < 0 ? y1 = 0 : y1 = y
    @dlg.set_position( x1, y1 ) if x1 != x || y1 != y
  end

  if !defined?(@loaded)
    UI.menu('Developer').add_item('CEF Version...') { self.cef_ver() }
    @loaded = true
    self.cef_ver(content_sizing: false)
  end

end
```

**From the testing it appears that the border width is always calculated at 8 pixels and the caption bar at 31 pixels.**
I am wondering if this has to do with old MFC window metrics no longer applying to Windows 10 with it's default single pixel window borders ?

?

**Comment by Eneroth3** [comment]
Logged as SKEXT-3147 and SKEXT-3148.

Thanks for reporting this!

**Comment by DanRathbun** [comment]
@thomthom Please review. Were the issues described above for `:use_content_size` were fixed in v2022.0 ?
